src/android/Dgram.java

```diff
@@ -75,198 +48,191 @@ public void run() {
+                final String address = data.getString(1);
+                final int sendPort = data.getInt(2);
+                cordova.getThreadPool().execute(new DatagramSocketSend(callbackContext,
+                        message,
+                        address,
+                        sendPort
+                ));
+                break;
+
+            case CLOSE_ACTION:
+                closeSocket();
+                callbackContext.success();
+                break;
+
+            default:
+                return false;
+        }
+
+        return true;
+    }
+
+    private void openSocket(String action, JSONArray data, CallbackContext callbackContext) throws JSONException {
+        closeSocket();
+        final int port = data.getInt(0);
+        final boolean isBroadcast = data.getBoolean(1);
+        try {
+            open(port, isBroadcast);
+            callbackContext.success();
+        } catch (SocketException e) {
+            Log.e(TAG, "Attempting " + action + " failed with: " + e.toString(), e);
+            callbackContext.error("'" + e.toString() + "'");
+        }
+    }
+
+    private void open(final int port, final boolean isBroadcast) throws SocketException {
         DatagramSocket datagramSocket = new DatagramSocket(port);
         datagramSocket.setBroadcast(isBroadcast);
-        m_datagramSockets.put(datagramSocketId, datagramSocket);
+        this.datagramSocket = datagramSocket;
     }
 
-    private void startListening(
-            final int datagramSocketId,
-            final DatagramSocket datagramSocket
-    ) {
-        closeListener(datagramSocketId);
+    private void startListening(CallbackContext callbackContext) {
+        closeListener();
+        onMessageCallback = callbackContext;
         DatagramSocketListener datagramSocketListener = new DatagramSocketListener(
-                datagramSocketId,
-                datagramSocket
         );
-        m_datagramSocketListeners.put(datagramSocketId, datagramSocketListener);
+        this.datagramSocketListener = datagramSocketListener;
         datagramSocketListener.start();
     }
 
-    private void close(
-            final int datagramSocketId,
-            final DatagramSocket datagramSocket
-    ) {
+    private void closeSocket() {
         if (datagramSocket != null) {
             if (!datagramSocket.isClosed()) {
                 datagramSocket.close();
             }
 
-            m_datagramSockets.remove(datagramSocketId);
-            closeListener(datagramSocketId);
+            this.datagramSocket = null;
+            closeListener();
         }
     }
 
-    private void closeListener(
-            final int datagramSocketId
-    ) {
-        final DatagramSocketListener datagramSocketListener = m_datagramSocketListeners.get(datagramSocketId);
+    private void closeListener() {
+        final DatagramSocketListener datagramSocketListener = this.datagramSocketListener;
```

Comment:
do we need to assign `this.datagramSocketListener` to a local `datagramSocketListener`?